src/util/id.rs

```diff
@@ -1,33 +1,311 @@
+/// Utilities for globally unique identifiers.
+///
+/// A UUID is made up of 2 parts. A global id and a local id. The global id acts as a identifier for
+/// local address spaces allowing systems to create their own generation methods for local ids while
+/// retaining global uniqueness.
+
+use std::cmp::Ordering;
 use std::collections::hash_map::DefaultHasher;
+use std::fmt::{Debug, Formatter};
 use std::hash::{Hash, Hasher};
+use std::num::NonZeroU64;
+use std::sync::Arc;
 
-/// Quickly identify and compare entities while retaining a human readable name.
+/// A global id backed by a 64 bit value.
+///
+/// Global ids are guaranteed to be globally unique. They are generated by an incrementing u64 bit
+/// internal counter and will always be non zero.
+///
+/// # Examples
+///
+/// ```
+/// use rosella_rs::util::id::GlobalId;
+///
+/// // Creates a new global id
+/// let id = GlobalId::new();
 ///
-/// comparing existing ID's is very fast so it is highly
-/// recommended to avoid creating new instances when not necessary. (Also reduces typing mistakes)
-#[derive(Clone, Debug, Eq)]
-pub struct NamedID {
-    pub name: String,
-    pub(crate) id: u64,
+/// // This is still the same id
+/// let same_id = id.clone();
+///
+/// assert_eq!(id, same_id);
+///
+/// // Creates a new different global id
+/// let other_id = GlobalId::new();
+///
+/// assert_ne!(id, other_id);
+///
+/// // 0 is a niche so Options are free
+/// assert_eq!(8, std::mem::size_of::<Option<GlobalId>>());
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct GlobalId(NonZeroU64);
+
+impl GlobalId {
+    // Need to reserve value 1 for the NamedId address space
+    const NEXT_GLOBAL_ID: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(2u64);
```

Comment:
`use` it?

Reply:
This is the only place where AtomicU64 is needed so a use seemed excessive.